### ISSUE TITLE
[JENKINS-40510] blueocean breaks when too many branches are used in parallel

### ISSUE THREAD

**Issue by jenkins-infra-bot** [jenkinsci/blueocean-plugin#3547]
<pre>
#!/usr/bin/env groovy

stage(<span class="code-quote">"Setup"</span>) {
}

stage(<span class="code-quote">"Tests"</span>) {
    def tests = [:]
    <span class="code-keyword">for</span> (<span class="code-object">int</span> i=0; i < 80; i++) {
def counter = i 
tests[i] = {
    node {
echo <span class="code-quote">"This is ${i}"</span>
    }   
}
    }

    parallel (tests)
}

stage(<span class="code-quote">"End"</span>) {
}
</pre>

<p>The above code is correctly displayed in blueocean in the form of:</p>

```
Setup -- Tests -- End
   -- 1
   -- ...
   -- 80

```

<p>However, if I increase the max count of the loop iterations from 80 to 100, then blueocean is not able to show the graph. Instead, I get the message:</p>

```
Error rendering PipelineRunGraph: TypeError: child is undefined

```

---
<details><summary><i>Originally reported by <a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=cgundogan">cgundogan</a>, imported from: <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-40510" target="_blank">blueocean breaks when too many branches are used in parallel</a></i></summary>
<i><ul>
<li><b>status</b>: Closed
<li><b>priority</b>: Major
<li><b>component(s)</b>: blueocean-plugin
<li><b>label(s)</b>: blueocean, pipeline
<li><b>resolution</b>: Duplicate
<li><b>resolved</b>: 2016-12-17T22:44:19+00:00
<li><b>votes</b>: 0
<li><b>watchers</b>: 2
<li><b>imported</b>: 2025-12-01
</ul></i>
<details><summary>Raw content of original issue</summary>

<pre>
<div class="code panel" style="border-width: 1px;"><div class="codeContent panelContent">
<pre class="code-java">#!/usr/bin/env groovy

stage(<span class="code-quote">"Setup"</span>) {
}

stage(<span class="code-quote">"Tests"</span>) {
    def tests = [:]
    <span class="code-keyword">for</span> (<span class="code-object">int</span> i=0; i &lt; 80; i++) {
        def counter = i 
        tests[i] = {
            node {
                echo <span class="code-quote">"This is ${i}"</span>
            }   
        }
    }

    parallel (tests)
}

stage(<span class="code-quote">"End"</span>) {
}
</pre>
</div></div>

<p>The above code is correctly displayed in blueocean in the form of:</p>

<div class="preformatted panel" style="border-width: 1px;"><div class="preformattedContent panelContent">
<pre>Setup -- Tests -- End
           -- 1
           -- ...
           -- 80
</pre>
</div></div>

<p>However, if I increase the max count of the loop iterations from 80 to 100, then blueocean is not able to show the graph. Instead, I get the message:</p>
<div class="preformatted panel" style="border-width: 1px;"><div class="preformattedContent panelContent">
<pre>Error rendering PipelineRunGraph: TypeError: child is undefined
</pre>
</div></div>
</pre>
</details>
</details>
<details><summary><i>environment</i></summary>

```
Setup includes multiple (&gt;10) workers.<br/>
Tested with<br/>
* Firefox 50.1.0<br/>
* BlueOcean beta 1.0.0-b13
```
</details>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/jamesdumay.png" title="jamesdumay's avatar" /> jamesdumay:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-40510?focusedId=280715&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-280715">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>Thanks for reporting this <a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=cgundogan" class="user-hover" rel="cgundogan">cgundogan</a>. This ticket appears to be a duplicate of <a href="https://issues.jenkins.io/browse/JENKINS-39770" title="Pipeline visualization not rendered when there is more that 100 nodes" class="issue-link" data-issue-key="JENKINS-39770">JENKINS-39770</a> so please watch and vote for it instead.</p>

<p>Question about the usage here - do you actually have a pipeline with 80+ parallels or are you just seeing what Blue Ocean can handle? </p>

<p>ping <a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=michaelneale" class="user-hover" rel="michaelneale">michaelneale</a> <a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=vivek" class="user-hover" rel="vivek">vivek</a></p>
</pre>
</details>

</ul>
</details>

<p>Thanks for reporting this <a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=cgundogan" class="user-hover" rel="cgundogan">cgundogan</a>. This ticket appears to be a duplicate of <a href="https://issues.jenkins.io/browse/JENKINS-39770" title="Pipeline visualization not rendered when there is more that 100 nodes" class="issue-link" data-issue-key="JENKINS-39770">JENKINS-39770</a> so please watch and vote for it instead.</p>

<p>Question about the usage here - do you actually have a pipeline with 80+ parallels or are you just seeing what Blue Ocean can handle? </p>

<p>ping <a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=michaelneale" class="user-hover" rel="michaelneale">michaelneale</a> <a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=vivek" class="user-hover" rel="vivek">vivek</a></p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i>cgundogan:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-40510?focusedId=280745&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-280745">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>Thank you <a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=jamesdumay" class="user-hover" rel="jamesdumay">jamesdumay</a> for the hint.</p>

<p>Indeed, I have a use case that requires &gt;100 nodes in the graph. My given example is highly simplified; I actually have a use case where I use two consecutive parallels that each have ~60 branches. Together, they result in many nodes at the UI side, hence breaking the graph.</p>
</pre>
</details>

</ul>
</details>

<p>Thank you <a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=jamesdumay" class="user-hover" rel="jamesdumay">jamesdumay</a> for the hint.</p>

<p>Indeed, I have a use case that requires >100 nodes in the graph. My given example is highly simplified; I actually have a use case where I use two consecutive parallels that each have ~60 branches. Together, they result in many nodes at the UI side, hence breaking the graph.</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/jamesdumay.png" title="jamesdumay's avatar" /> jamesdumay:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-40510?focusedId=280759&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-280759">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=cgundogan" class="user-hover" rel="cgundogan">cgundogan</a> thats good to know. We may not be able to make the Pipeline graph scale to that amount of parallels but I'd love to chat with you about this. Would you mind having a quick Google hangout to chat about your specific usage?</p>
</pre>
</details>

</ul>
</details>

<p><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=cgundogan" class="user-hover" rel="cgundogan">cgundogan</a> thats good to know. We may not be able to make the Pipeline graph scale to that amount of parallels but I'd love to chat with you about this. Would you mind having a quick Google hangout to chat about your specific usage?</p>

**Comment by jenkins-infra-bot** [comment]
<i>[Original `duplicates` from Jira: <a href="https://github.com/jenkinsci/blueocean-plugin/issues?q=is%3Aissue%20%22jira_issue_key%3DJENKINS-39770%22">JENKINS-39770</a>]</i>
